Common/CUDA/tvdenoising.cu

```diff
@@ -399,48 +412,55 @@ do { \
                     }
-                    for (dev = 0; dev < deviceCount; dev++){   
+                    for (dev = 0; dev < deviceCount; dev++){ 
+                        is_last_chunk=!((sp*deviceCount+dev)<deviceCount*splits-1);
+                        is_first_chunk=!(sp*deviceCount+dev);
                         cudaSetDevice(gpuids[dev]);
                         cudaStreamSynchronize(stream[dev*nStream_device+3]);
-                        // Sometimes the last_chunk is smaller than the other ones, and thus if we don't reset the memory to zero, we'll gave ghosts residual data in the variable
-                        cudaMemsetAsync(d_py[dev], 0, mem_img_each_GPU,stream[dev*nStream_device+1]); 
-                        cudaMemcpyAsync(d_py[dev] +offset_device[dev], h_py+offset_host[dev],  bytes_device[dev]*sizeof(float), cudaMemcpyHostToDevice,stream[dev*nStream_device+3]);
-                        
+                        if (is_last_chunk) {cudaMemsetAsync(d_py[dev], 0, mem_img_each_GPU,stream[dev*nStream_device+3]);}
+                        cpy_from_host(d_py[dev],h_py,bytes_device[dev], offset_device[dev],offset_host[dev], pixels_per_slice, buffer_length, stream[dev*nStream_device+3],  is_first_chunk,  is_last_chunk, image_size);
                     }
-                    for (dev = 0; dev < deviceCount; dev++){   
+                    for (dev = 0; dev < deviceCount; dev++){ 
+                        is_last_chunk=!((sp*deviceCount+dev)<deviceCount*splits-1);
+                        is_first_chunk=!(sp*deviceCount+dev);
                         cudaSetDevice(gpuids[dev]);
                         cudaStreamSynchronize(stream[dev*nStream_device+4]);
-                        // Sometimes the last_chunk is smaller than the other ones, and thus if we don't reset the memory to zero, we'll gave ghosts residual data in the variable
-                        cudaMemsetAsync(d_pz[dev], 0, mem_img_each_GPU,stream[dev*nStream_device+1]); 
-                        cudaMemcpyAsync(d_pz[dev] +offset_device[dev], h_pz+offset_host[dev],  bytes_device[dev]*sizeof(float), cudaMemcpyHostToDevice,stream[dev*nStream_device+4]);
-                        
-                    } 
-                    for (dev = 0; dev < deviceCount; dev++){   
+                        if (is_last_chunk) {cudaMemsetAsync(d_pz[dev], 0, mem_img_each_GPU,stream[dev*nStream_device+4]);}
+                        cpy_from_host(d_pz[dev],h_pz,bytes_device[dev], offset_device[dev],offset_host[dev], pixels_per_slice, buffer_length, stream[dev*nStream_device+4],  is_first_chunk,  is_last_chunk, image_size);
+                        // Z derivative must be negated in sign to keep Neumman conditions
+                        if (is_first_chunk){
+                            multiplyArrayScalar<<<60,MAXTREADS,0,stream[dev*nStream_device+4]>>>(d_pz[dev],             -1,  pixels_per_slice*bytes_device[dev]);    

```

Comment:
there is a mistake, it is not 'bytes_device', it is 'buffer_length'

Reply:
```suggestion
                            multiplyArrayScalar<<<60,MAXTREADS,0,stream[dev*nStream_device+4]>>>(d_pz[dev],             -1,  pixels_per_slice*bytes_device[dev]);    
```
youu are right, apologies, Im quite distracted these days!